src/shared/mentionable_text_input.rs

```diff
@@ -495,87 +1024,480 @@ impl MentionableTextInput {
 
 impl MentionableTextInputRef {
     pub fn text(&self) -> String {
         self.borrow().map_or_else(String::new, |inner| inner.text())
     }
 
+    /// Returns a reference to the inner `TextInput` widget.
+    pub fn text_input_ref(&self) -> TextInputRef {
+        self.borrow()
+            .map(|inner| inner.cmd_text_input.text_input_ref())
+            .unwrap_or_default()
+    }
+
     pub fn set_text(&self, cx: &mut Cx, text: &str) {
         if let Some(mut inner) = self.borrow_mut() {
             inner.set_text(cx, text);
         }
     }
 
-    /// Returns a reference to the inner `TextInput` widget.
-    pub fn text_input_ref(&self) -> TextInputRef {
-        self.borrow()
-            .map(|inner| inner.cmd_text_input.text_input_ref())
-            .unwrap_or_default()
+    /// Extracts existing @room and user mentions from the current text
+    pub fn extract_existing_mentions(&self, cx: &mut Cx) {
+        if let Some(mut inner) = self.borrow_mut() {
+            inner.extract_existing_mentions(cx);
+        }
     }
 
-    /// Sets the room members for this text input
-    pub fn set_room_members(&self, members: Arc<Vec<RoomMember>>) {
+    /// Extracts existing mentions from both the original Matrix event data and text analysis.
+    /// This method should be used when editing a message to preserve the original mentions.
+    ///
+    /// Priority:
+    /// 1. First tries to extract from the original Matrix event data (most reliable)
+    /// 2. Falls back to text analysis if event data is unavailable
+    pub fn extract_existing_mentions_from_event(&self, cx: &mut Cx, original_event_item: &matrix_sdk_ui::timeline::EventTimelineItem) {
         if let Some(mut inner) = self.borrow_mut() {
-            inner.set_room_members(members);
+            inner.extract_existing_mentions_with_event(cx, Some(original_event_item));
         }
     }
 
-    pub fn set_room_id(&self, room_id: OwnedRoomId) {
+    /// Sets whether the current user can notify the entire room (@room mention)
+    pub fn set_can_notify_room(&self, can_notify: bool) {
         if let Some(mut inner) = self.borrow_mut() {
-            inner.set_room_id(room_id);
+            inner.set_can_notify_room(can_notify);
         }
     }
 
-    pub fn get_room_id(&self) -> Option<OwnedRoomId> {
-        self.borrow().and_then(|inner| inner.get_room_id())
+    /// Gets whether the current user can notify the entire room (@room mention)
+    pub fn can_notify_room(&self) -> bool {
+        self.borrow().is_some_and(|inner| inner.can_notify_room())
     }
 
-    /// Returns the list of users mentioned in the given html message content.
-    pub fn get_real_mentions_in_html_text(&self, html: &str) -> BTreeSet<OwnedUserId> {
-        let Some(inner) = self.borrow() else { return BTreeSet::new() };
-        let mut real_mentions = BTreeSet::new();
+
+    /// Returns the mentions analysis for the given html message content.
+    /// Returns (user_mentions, has_room_mention)
```

Comment:
what is "mentions analysis"? that's vague. Plus the second line is now outdated.

```suggestion
    /// Returns the mentions actually present in the given html message content.
```